apiserver/facades/controller/crossmodelrelations/relationunitswatcher.go

```diff
@@ -0,0 +1,187 @@
+		offerApplicationUUID: offerApplicationUUID,
+		offerRelationUUID:    offerRelationUUID,
+		service:              service,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Name: "relation-units-watcher",
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{source},
+	})
+	return w, errors.Capture(err)
+}
+
+type relationChangesWatcher struct {
+	source   watcher.NotifyWatcher
+	out      chan params.RelationUnitsChange
+	catacomb catacomb.Catacomb
+	service  RelationUnitsWatcherService
+
+	offerApplicationUUID coreapplication.UUID
+	offerRelationUUID    corerelation.UUID
+	data                 relation.ConsumerRelationUnitsChange
+}
+
+func (w *relationChangesWatcher) loop() error {
+	ctx := w.catacomb.Context(context.Background())
+
+	data, err := w.service.GetConsumerRelationUnitsChange(ctx, w.offerRelationUUID, w.offerApplicationUUID)
+	if err != nil {
+		return errors.Errorf("fetching consumer side of relation %v: %w", w.offerRelationUUID, err)
+	}
+	w.data = data
+
+	// We need to close the changes channel because we're inside the
+	// API - see apiserver/watcher.go:srvRelationUnitsWatcher.Next()
+	defer close(w.out)
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case _, ok := <-w.source.Changes():
+			if !ok {
+				select {
+				case <-w.catacomb.Dying():
+					return w.catacomb.ErrDying()
+				default:
+					return errors.New("relation units watcher closed")
+				}
+			}
+
+			unitRelationInfo, err := w.service.GetConsumerRelationUnitsChange(ctx, w.offerRelationUUID, w.offerApplicationUUID)
+			if err != nil {
+				return errors.Errorf("fetching consumer side of relation %v: %w", w.offerRelationUUID, err)
+			}
+
+			info, sendEvent := w.convert(unitRelationInfo)
+			if !sendEvent {
+				continue
+			}
+
+			w.data = unitRelationInfo
+
+			// Send in lockstep so we don't drop events.
+			select {
+			case <-w.catacomb.Dying():
+				return w.catacomb.ErrDying()
+			case w.out <- info:
+			}
+		}
+	}
+}
+
+func (w *relationChangesWatcher) convert(
+	newChange relation.ConsumerRelationUnitsChange,
+) (params.RelationUnitsChange, bool) {
+	changes := w.dataChanges(newChange)
+	if changes.Empty() {
+		return params.RelationUnitsChange{}, false
+	}
+
+	unitsChanged := transform.Map(changes.UnitsSettingsVersions, func(key string, val int64) (string, params.UnitSettings) {
+		return key, params.UnitSettings{Version: val}
+	})
+
+	return params.RelationUnitsChange{
+		Changed:    unitsChanged,
+		AppChanged: changes.AppSettingsVersion,
+		Departed:   changes.DepartedUnits,
+	}, true
+}
+
+// dataChanges returns the changed data since the last time this watcher
+// was triggered.
+func (w *relationChangesWatcher) dataChanges(event relation.ConsumerRelationUnitsChange) relation.ConsumerRelationUnitsChange {
+	changedEvent := relation.ConsumerRelationUnitsChange{}
+
+	departed := set.NewStrings(event.DepartedUnits...).Difference(set.NewStrings(w.data.DepartedUnits...))
```

Comment:
This is left aligned. So it'll only tell you if have the following, then it'll show nothing:

```go
a := set.NewStrings("a", "b").Difference(set.NewStrings("a", "b", "c"))
// map[]
```

Which I think is ok here, you only want to know what was different from the last event.